### PR TITLE
Changes in TaQL grammar for new commands

### PR DESCRIPTION
**Pull request by gervandiepen**
This is the second part of the issue #1143 changes.
It contains the changes in the TaQL grammar and the 1st stage parsing. The actual implementation of the new commands is part of another PR.

### CODE REVIEW

**Review by mordante — COMMENTED** [review]
SGTM in general but several minor nits.

File: tables/TaQL/TaQLNodeDer.cc [hunk 1]
```diff
@@ -1589,8 +1591,23 @@ void TaQLCreTabNodeRep::showDerived (std::ostream& os) const
...
+    os << " LIKE ";
+    const std::vector<TaQLNode>& nodes = itsLikeDrop.getMultiRep()->getNodes();
+    nodes[0].show (os);
+    if (nodes.size() == 2) {
+      os << " DROP COLUMN ";
+      nodes[1].show (os);
+    }
+  }
+  if (itsColumns.isValid() && !itsColumns.getMultiRep()->getNodes().empty()) {
+    if (itsLikeDrop.isValid()) {
```
[the remark this answers is not in the document]
Comment (gervandiepen) [reply tables/TaQL/TaQLNodeDer.cc:1604]: Added comments

File: tables/TaQL/TaQLNodeDer.cc [hunk 2]
```diff
@@ -1613,22 +1631,25 @@ TaQLCreTabNodeRep* TaQLCreTabNodeRep::restore (AipsIO& aio)
...
-    itsDtype (checkDataType(dtype)),
-    itsSpec  (spec)
+  : TaQLNodeRep  (TaQLNode_ColSpec),
+    itsName    (name),
+    itsLikeCol (likeCol),
+    itsDtype   (checkDataType(dtype)),
+    itsSpec    (spec)
 {}
 TaQLColSpecNodeRep::~TaQLColSpecNodeRep()
 {}
```
[the remark this answers is not in the document]
Comment (gervandiepen) [reply tables/TaQL/TaQLNodeDer.cc:1655]: Removed all 'empty' destructors

File: tables/apps/taql.cc [hunk 3]
```diff
@@ -267,11 +267,14 @@ Bool readLine (String& line, const String& prompt)
...
 
-// Read a line until a non-empty line is read. 
+// Read a line until a non-empty line or ^D or quit is read. 
 Bool readLineSkip (String& line, const String& prompt)
 {
   Bool fnd = False;
   while (!fnd  &&  readLine (line, prompt)) {
+    if (line=="exit" || line=="quit" || line=="q") {
+      break;
+    }
```
[the remark this answers is not in the document]
Comment (gervandiepen) [reply tables/apps/taql.cc:277]: Removed

File: tables/TaQL/TableGram.ll [hunk 4]
```diff
@@ -337,11 +341,21 @@ PATTREX   {OPERREX}{WHITE}({PATTEX}|{DISTEX})
 	    BEGIN(TABLENAMEstate);
 	    return ALTERTAB;
 	  }
+{DROPTAB} {
+            tableGramPosition() += yyleng;
+	    BEGIN(TABLENAMEstate);
```
[the remark this answers is not in the document]
Comment (gervandiepen) [reply tables/TaQL/TableGram.ll:346]: Removed all tabs by spaces

File: tables/TaQL/TaQLNodeDer.cc [hunk 5]
```diff
@@ -1613,22 +1631,25 @@ TaQLCreTabNodeRep* TaQLCreTabNodeRep::restore (AipsIO& aio)
...
   TaQLMultiNode with = TaQLNode::restoreMultiNode (aio);
   TaQLNode giving = TaQLNode::restoreNode (aio);
+  TaQLMultiNode likeDrop = TaQLNode::restoreMultiNode (aio);
   TaQLMultiNode columns = TaQLNode::restoreMultiNode (aio);
   TaQLNode limit = TaQLNode::restoreNode (aio);
   TaQLMultiNode dminfo = TaQLNode::restoreMultiNode (aio);
-  TaQLCreTabNodeRep* node = new TaQLCreTabNodeRep (with, giving, columns,
-                                                   limit, dminfo);
+  TaQLCreTabNodeRep* node = new TaQLCreTabNodeRep (with, giving, likeDrop,
+                                                   columns, limit, dminfo);
```
[the remark this answers is not in the document]
Comment (gervandiepen) [reply tables/TaQL/TaQLNodeDer.cc:1639]: Use std::unique_ptr where needed.
Note that result of restore function is put in a TaQLNose object which is exception-safe.

File: tables/TaQL/TaQLNodeDer.cc [hunk 6]
```diff
@@ -1649,15 +1673,15 @@ void TaQLColSpecNodeRep::show (std::ostream& os) const
...
 {
-  aio << itsName << itsDtype;
+  aio << itsName << itsLikeCol << itsDtype;
   itsSpec.saveNode (aio);
 }
 TaQLColSpecNodeRep* TaQLColSpecNodeRep::restore (AipsIO& aio)
 {
-  String name, dtype;
-  aio >> name >> dtype;
+  String name, likeCol, dtype;
```
[the remark this answers is not in the document]
Comment (gervandiepen) [reply tables/TaQL/TaQLNodeDer.cc:1681]: There is no Casacore style guide. There was an old Programming Guide which is obsolete now.

What exactly do you mean with consistent naming scheme? The names used are consistent with the class member names.

Why one variable per line? If they are initialized I fully agree, but only a declaration looks fine to me. Note that the line thereafter (with shift operators) also contains multiple variables.

File: tables/TaQL/TaQLNodeDer.cc (hunk 1 above)
[the remark this answers is not in the document]
Comment (mordante) [reply tables/TaQL/TaQLNodeDer.cc:1604]: Did you (force) push your changes?

**Comment by gervandiepen** [comment]
Not yet. I was waiting for the answer to that one question.


Op di 22 feb. 2022 om 08:24 schreef mordante ***@***.***>

File: tables/TaQL/TaQLNodeDer.cc (hunk 6 above)
[the remark this answers is not in the document]
Comment (mordante) [reply tables/TaQL/TaQLNodeDer.cc:1681]: I looked a bit further and this style seems the names like `likeCol` aren't too uncommon. It seems the `_` is only used for a Hungary naming scheme.

File: tables/TaQL/TaQLNodeDer.cc [hunk 7]
```diff
@@ -107,8 +107,6 @@ TaQLConstNodeRep::TaQLConstNodeRep (Int64 value, const String& subTableName)
     itsCValue      (value, 0.),
     itsSValue      (subTableName)
 {}
-TaQLConstNodeRep::~TaQLConstNodeRep()
```
Comment (mordante) [inline tables/TaQL/TaQLNodeDer.cc:110]: Next time please make a separate PR for cleaning up existing code. When this PR introduces a bug it's a lot less code to review to determine the bug.
I much rather review 2 PRs, one cleanup and one code fixes.

File: tables/TaQL/TaQLNodeDer.cc (hunk 5 above)
[the remark this answers is not in the document]
Comment (mordante) [reply tables/TaQL/TaQLNodeDer.cc]: Even when they don't throw now they may throw in the feature. I'm not fond of using raw owning pointers in C++11, since they tend to lead to bugs. Since casacore originated before C++11 it doesn't use the new smart pointers a lot. I still think it would be good to look at using them more. Especially in code that doesn't affect the API.

**Comment by mordante** [comment]
LGTM! I leave it to you discretion whether you want to split the NFC changes to a separate commit.

File: tables/TaQL/TaQLNodeDer.cc (hunk 7 above)
Comment (gervandiepen) [reply tables/TaQL/TaQLNodeDer.cc:110]: Undid superfluous changes (I assume NFC means non-functional changes).

Comment (mordante) [reply tables/TaQL/TaQLNodeDer.cc:110]: NFC indeed is non-functional change.
